devito/mpi/distributed.py

```diff
@@ -600,17 +616,53 @@ class CustomTopology(tuple):
     """
 
     def __new__(cls, items, input_comm):
+        # Keep track of nstars and already defined decompositions
         nstars = len([i for i in items if i == '*'])
-        if nstars > 0:
-            if input_comm.size % nstars != 0:
-                raise ValueError("Invalid `topology` for given nprocs")
-            if any(i not in ('*', 1) for i in items):
-                raise ValueError("Custom topology must be only 1 or *")
-
-            v = input_comm.size // nstars
-            processed = [i if i == 1 else v for i in items]
-        else:
+
+        # If no stars exist we are ready
+        if nstars == 0:
             processed = items
+        else:
+            # Init decomposition list
+            processed = [1] * len(items)
+
+            # Get star and integer indices
+            int_pos = [i for i, item in enumerate(items) if isinstance(item, int)]
+            int_vals = [item for item in items if isinstance(item, int)]
+            star_pos = [i for i, item in enumerate(items) if not isinstance(item, int)]
+
+            # Decompose the processes remaining for allocation to prime factors
+            alloc_procs = np.prod([i for i in items if i != '*'])
+            remprocs = int(input_comm.size // alloc_procs)
+            prime_factors = primefactors(remprocs)
```

Comment:
instead of having this both here and inside the while loop, you should have it only once, at the top of the while loop...

which, if necessary, you can turn into:

```
while True:
    if rem_procs > 1:
        break
    ...
```

thus emulating a do-while loop

Reply:
Dropped and improved, lemme know if it is fine